File: packages/contracts/src/catalog.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

export const SCHEMA_VERSION = 1;
export const SCHEMA_ID = "evidenceops.contracts.v1";

export type Scope = "none" | "tenant" | "tenant_matter";

export type NodeTypeDef = {
  name: string;
  prefix: string;
  scope: Scope;
  naturalKey: string[];
};

export type Catalog = {
  schemaVersion: number;
  schemaId: string;
  verdicts: string[];
  reviewerStates: string[];
  redactionClassifications: string[];
  dataAvailability: string[];
  verdictMethods: string[];
  impactKinds: string[];
  reasonCodes: string[];
  nodeTypes: NodeTypeDef[];
  edgeTypes: string[];
};

let cached: Catalog | undefined;

export function catalogPath(): string {
  const here = dirname(fileURLToPath(import.meta.url));
  return join(here, "../schemas/v1/catalog.json");
}

export function loadCatalog(): Catalog {
  if (cached) {
    return cached;
  }
  const parsed: unknown = JSON.parse(readFileSync(catalogPath(), "utf8"));
  if (typeof parsed !== "object" || parsed === null) {
    throw new Error("catalog: invalid JSON");
  }
  cached = parsed as Catalog;
  if (cached.schemaVersion !== SCHEMA_VERSION) {
    throw new Error("catalog: unexpected schemaVersion");
  }
  return cached;
}

export function nodeTypeDef(name: string): NodeTypeDef {
  const def = loadCatalog().nodeTypes.find((item) => item.name === name);
  if (!def) {
    throw new Error(`unknown node type: ${name}`);
  }
  return def;
}

File: src/evidenceops/contracts/catalog.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any, TypedDict

SCHEMA_VERSION = 1
SCHEMA_ID = "evidenceops.contracts.v1"


class NodeTypeDef(TypedDict):
    name: str
    prefix: str
    scope: str
    naturalKey: list[str]


class Catalog(TypedDict):
    schemaVersion: int
    schemaId: str
    verdicts: list[str]
    reviewerStates: list[str]
    redactionClassifications: list[str]
    dataAvailability: list[str]
    verdictMethods: list[str]
    impactKinds: list[str]
    reasonCodes: list[str]
    nodeTypes: list[NodeTypeDef]
    edgeTypes: list[str]


def catalog_path() -> Path:
    return (
        Path(__file__).resolve().parents[3]
        / "packages"
        / "contracts"
        / "schemas"
        / "v1"
        / "catalog.json"
    )


@lru_cache(maxsize=1)
def load_catalog() -> Catalog:
    parsed: Any = json.loads(catalog_path().read_text(encoding="utf-8"))
    if parsed.get("schemaVersion") != SCHEMA_VERSION:
        raise ValueError("catalog: unexpected schemaVersion")
    return parsed


def node_type_def(name: str) -> NodeTypeDef:
    for item in load_catalog()["nodeTypes"]:
        if item["name"] == name:
            return item
    raise KeyError(f"unknown node type: {name}")
